### ISSUE TITLE
Inconsistent data on phone versus server

### ISSUE THREAD

**Issue by bardram** [cph-cachet/carp-webservices-spring#16]
During data sampling over 24 hours, data was collected on the phone and uploaded (streamed) to CAWS. Data was also stored on the phone (using the SQLite DB). When later downloading the data from the server (as JSON) and comparing, there is SIGNIFICANT less data on the server as compared to the phone. Here is the comparison table:

![image](https://github.com/cph-cachet/carp-webservices-spring/assets/1196642/0029914f-ca86-4dff-a502-685af181389f)

As can be seen, there is a huge difference - there are practically no data downloaded from the server. 

There are three potential root causes

1.  Data is not uploaded - should be checked by @bardram 
2. Data is not stored in the CAWS DB - should be checked by @xelahalo 
3. Data is not exported correctly - should be checked by @xelahalo 

The data and Jupyter notebook is available [here](https://github.com/cph-cachet/carp.analysis.sandbox/tree/master/projects/carp.mobile.sensing/comparison).

**Comment by bardram** [comment]
After some investigations, I found that data from connected devices (e.g. HR) is not streamed properly. Will fix this first and re-run the test.

**Comment by bardram** [comment]
When trying to upload the data from a study based on the protocol below, I get 

```
CarpServiceException: 500 Internal Server Error -  No value present - POST /api/data-stream-service
```

The underlying status response is 

````
RESPONSE: 500
      {"statusCode":500,"exception":"java.util.NoSuchElementException","message":"No value present","path":"POST /api/data-stream-service"}
````

What does this mean?

<details>
  <summary>Data</summary>
  
  ```
    [Copy-paste relevant console print-out here]
  ```

</details>

**Comment by bardram** [comment]
Trying to upload another batch of data, I get

```
 {"statusCode":500,"exception":"java.util.NoSuchElementException","message":"No value present","path":"POST /api/data-stream-service"}
````

Still same error - missing `path` element - but where?

**Comment by bardram** [comment]
Trying to upload a very simple batch of data below, but still gets this exception.

``` 
{
    "__type": "dk.cachet.carp.data.infrastructure.DataStreamServiceRequest.AppendToDataStreams",
    "apiVersion": "1.1",
    "studyDeploymentId": "3461376a-dbd9-49c3-a3e5-16ebfb844160",
    "batch": [
        {
            "dataStream": {
                "studyDeploymentId": "3461376a-dbd9-49c3-a3e5-16ebfb844160",
                "deviceRoleName": "Primary Phone",
                "dataType": "dk.cachet.carp.freememory"
            },
            "firstSequenceId": 1,
            "measurements": [
                {
                    "sensorStartTime": 1710762458387416,
                    "data": {
                        "__type": "dk.cachet.carp.freememory",
                        "freePhysicalMemory": 137916416,
                        "freeVirtualMemory": 2945175552
                    }
                },
                {
                    "sensorStartTime": 1710762518390908,
                    "data": {
                        "__type": "dk.cachet.carp.freememory",
                        "freePhysicalMemory": 143290368,
                        "freeVirtualMemory": 2949804032
                    }
                },
                {
                    "sensorStartTime": 1710762578423843,
                    "data": {
                        "__type": "dk.cachet.carp.freememory",
                        "freePhysicalMemory": 630353920,
                        "freeVirtualMemory": 3438329856
                    }
                }
            ],
            "triggerIds": [
                1
            ]
        }
    ]
}
```

**Comment by bardram** [comment]
Now - after the new deployment, this exception is not happening anymore, and I can upload data 🎉

**Comment by bardram** [comment]
After re-running my test with some very simple measures, I get the following results:

![image](https://github.com/cph-cachet/carp-webservices-spring/assets/1196642/1c7661fc-01bd-4568-9712-28074c128bd9)

Now the problem seems to be reversed - there are way more measurements downloaded as JSON from the server that actually sampled on the phone and stored in the SQLite DB??????

When looking into the JSON each data stream (which has a unique id) is present multiple times, instead of one. For example `"data_stream_id": 212` (which contains some `freememory` measurements) are in the JSON 1809 times!

The data is available [here]().

So - there might be two problems:

 * the data (measurements) are stored multiple times
 * the script extracting the json act up

**Comment by davidscavnicky** [comment]
I have checked work done by me - https://github.com/cph-cachet/carp-webservices-spring/commit/bda12b2c4eb5de31c37edd8be6c79d0ed38e3d5d . 

On the backend side I tried to generate summary and compare with the same summary of a study you added as data. but unsure about studyID + is carp-data.db up to date when the data was generated? Could be that the .db is missing data, in the first place. @bardram 

I mean - there should some flow of comparing data, so backend is "respected" - study created, data uploaded from device, backend creates .json files in summary .zip, download.

the script extracting json is a function used throughout Webservices and there shouldn't be issue with matching snapshot attributes to data classes, but that also needs probably more investigation.

I think we need to narrow down the problem. and it would be super helpful to write some tests

**Comment by davidscavnicky** [comment]
I have connected to deployed db, and generated the summaries with he deployment id is 72b6b62c-1881-4df8-b3ee-c7651b6bc871.

the generated .json files are correct when compared with data [here](https://github.com/cph-cachet/carp.analysis.sandbox/tree/master/projects/carp.mobile.sensing/comparison/2024.03.19/70fb1912_2024-03-18T21_33_22.626299958Z) @bardram .

class createSummaryForStudy --> class ResourceExporterServiceImpl --> fun exportDataStreamSequences(...)

    @Query(nativeQuery = true,
            value = "SELECT dsi.id, ds.data_stream_id, ds.snapshot, ds.first_sequence_id, ds.last_sequence_id, ds.updated_at, ds.created_at, ds.created_by, ds.updated_by \n" +
                    "FROM data_stream_sequence ds\n" +
                    "left join data_stream_ids dsi on ds.data_stream_id = dsi.id\n" +
                    "WHERE dsi.study_deployment_id IN (:deploymentIds)")
    fun findAllByDeploymentIds(@Param("deploymentIds") deploymentIds: List<String>): List<DataStreamSequence>

-could be an issue the way to tables joint together, I will try to test different queries(inner) and talk with @xelahalo.

NOW,

data_stream_ids table:

![Screenshot 2024-04-09 at 22 41 11](https://github.com/cph-cachet/carp-webservices-spring/assets/109144925/c571dd95-cd6a-48cb-bd85-cd081190d9d3)

data_stream_sequence table:

![Screenshot 2024-04-09 at 22 42 46](https://github.com/cph-cachet/carp-webservices-spring/assets/109144925/4862f21a-08fa-461b-90ec-747444814158)

column created_at example :

2024-03-26 17:08:40.375045
2024-03-26 17:08:40.377596
2024-03-26 17:08:40.380139
2024-03-26 17:08:40.382634

many of the data are created at the same time practically and question could be how useful are they, but I can confirm that there is this amount of data in deployed database that corresponds to ratios(between JSON and Sqlite) you measured @bardram. Further investigation is required.

I lack understanding how phone stores data within 24 hours range(the way you measured) and what is stashed in SQlite and what's not. Could have some filtering of received data from sensor... that checks if that entry is not already there before uploading to deployed db, which then is taken from and generate .json files

**Comment by bardram** [comment]
It seems like the same data is added to the database multiple times but with 1 minute apart. Maybe all the data from the phone is accululated and uploaded multiple times?

@bardram - check what data is uploaded. Make a recording of the appendToDataStream calls so that this can be replayed.

**Comment by bardram** [comment]
Hi @davidscavnicky and @xelahalo  - I've now tested the above and I (still) don't think it is a problem with the upload of data. 

I ran a small test and you can see the results here:

https://github.com/cph-cachet/carp.analysis.sandbox/tree/master/projects/carp.mobile.sensing/comparison/2024.04.10

where the log shows that the phone uploads two batches, the first with 5 streams and 5 measurement and the second with 1 stream with 3 measurements.

So - there is no duplicate upload of data from the phone's side.

**Comment by davidscavnicky** [comment]
table data_stream_sequence contains duplicates with different created_at column entries (createdat is assigned when put into deployed db).

e.g. the id 212 containing freememory, with same first_sequence_id = 12 and last_sequence_id = 17 - shouldn't be allowed in a chain of uploads.

![Screenshot 2024-04-10 at 15 12 24](https://github.com/cph-cachet/carp-webservices-spring/assets/109144925/838e0dc7-f289-4348-aa74-da8a4a1f5bfd)

**Comment by bardram** [comment]
Seems like the data is accumulated in a buffer on the server and **ALL** accumulated data is added to the database over and over again every time the endpoint is called.

**Comment by davidscavnicky** [comment]
Here, it is shown in what manner the data was accumulating ( POST of two data streams, 4 times in a row)

![Screenshot 2024-04-13 at 15 14 39](https://github.com/cph-cachet/carp-webservices-spring/assets/109144925/611d5306-086f-425b-b9c6-9bdee8551ae5)

-----

https://github.com/cph-cachet/carp-webservices-spring/pull/32

I have made changes and tested functionality. I uploaded this data in 3 POSTs: https://github.com/cph-cachet/carp.analysis.sandbox/tree/master/projects/carp.mobile.sensing/comparison/2024.04.10 @bardram 

and checked that the number of data streams in .json files corresponds to data in db table data_stream_sequence.

**Comment by davidscavnicky** [comment]
Submitted a pull request: https://github.com/cph-cachet/carp-webservices-spring/pull/32

I have made changes and tested functionality. I uploaded this data in 3 POSTs: https://github.com/cph-cachet/carp.analysis.sandbox/tree/master/projects/carp.mobile.sensing/comparison/2024.04.10 @bardram 

and checked that the number of data streams in .json files corresponds to data in db table data_stream_sequence.

**Comment by bardram** [comment]
As part of the 3.0.0 release of the carp_webservices, I've made a simple unit test of this issue. The test uploads one batch and then downloads all batches, and then checks that the number of batches is increased with one (1) batch at a time. 

At the time of testing / writing, this unit test fails:

![image](https://github.com/cph-cachet/carp-webservices-spring/assets/1196642/513c960e-a633-4870-8acb-a4c5d140e8a1)

But - the PR above isn't released yet, as I can see, so at least we now have a unit test that can test for this...

**Comment by xelahalo** [comment]
Fixed with #32
